PHP 8.1 compatibility

agustingomes opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | no
| New Feature   | yes https://github.com/mezzio/mezzio/issues/95
| RFC           | no
| QA            | no

### Description

This PR brings compatibility with PHP 8.1

#### Notes

- all the direct dependencies under `require` are ready for PHP 8.1
- all the following dev dependencies need to be upgraded to be compatible with PHP 8.1
  - `mezzio/mezzio-aurarouter`
  - `mezzio/mezzio-laminasrouter`

froschdesign commented on composer.json:
```diff
@@ -30,7 +30,7 @@
     },
     "require": {
-        "php": "^7.3||~8.0.0",
+        "php": "^7.3||~8.0.0||~8.1.0",
```
Version 7.3 of PHP is dead in 11 days, so it can be removed.

yethee commented on phpcs.xml:
```diff
@@ -20,4 +20,14 @@
     <exclude-pattern>test/TestAsset/**</exclude-pattern>
     <exclude-pattern>test/**/TestAsset/**</exclude-pattern>
+
+    <!-- Needed to avoid breaking code for 7.4 since $instance::class only works in PHP 8.0 -->
```
```xml
<config name="php_version" value="70400" />
```
This option should solve issue of compat with PHP 7.4

boesing replied:
I don't get why we would need this in the first place. Coding Standard should only be executed on PHP 7.4, so no need to configure the version at all.

If you can't manage your local environment to run under PHP 7.4, then you might check how to achieve this instead of enforcing config parameters which might be overseen when bumping to PHP 8.0

boesing commented on phpcs.xml:
```diff
@@ -20,4 +20,14 @@
     <exclude-pattern>test/**/TestAsset/**</exclude-pattern>
+
+    <!-- Needed to avoid breaking code for 7.4 since $instance::class only works in PHP 8.0 -->
+    <rule ref="SlevomatCodingStandard.Classes.ModernClassNameReference.ClassNameReferencedViaFunctionCall">
```
For what are these necessary? Not sure if I understand this change, I think it has to do with the fact that you were running code-style checks with PHP > 8.0, so please ensure that you execute codestyle locally with the lowest supported PHP version of this component (which is actually PHP 7.4 as per `composer.json`.

agustingomes replied:
> I think it has to do with the fact that you were running code-style checks with PHP > 8.0

Ah that is indeed the case. completely missed to rebuild the images I use for development with 7.4 before performing these fixes. will review the second commit and try to minimize the changes. Thanks for pointing this out.

boesing commented:
I think creating a `.laminas-ci.json` where we allow ignoring platform dependencies for PHP 8.1.
https://github.com/laminas/laminas-ci-matrix-action#configuration

`aura` router uses `auryn` repository and afair, it has its latest commit in 2017 and thus, will be dropped in the next versions anyways. See #94 for more details.

agustingomes commented:
@boesing I added a mechanism to avoid testing the Aura router against PHP 8.1 and higher, let me know if that works for covering the appendix section of https://github.com/mezzio/mezzio/issues/94 

As for the static analysis, I guess they are breaking because some changes from the dependent packages, but I may need some help with them, as I'm not very familiarized with Psalm.

agustingomes committed:
Update packages to maximum PHP 8.1 compatibility

This allows new dependency versions to be locked and in practice
makes all the packages that are in the `require` section compatible
with PHP 8.1.

Only 2 dev-require packages remain incompatible with this commit:
```sh
$ composer why-not php 8.1
mezzio/mezzio-aurarouter     3.1.0  requires  php (^7.3||~8.0.0)
mezzio/mezzio-laminasrouter  3.1.0  requires  php (^7.3 || ~8.0.0)
```

Prevents the tests for the lowest PHP 8.1 dependencies to fail by
indicating conflicting packages that are not directly required in
this library, but that they fail in some form under PHP 8.1.

For the Aura route package, the tests will exclude that adapter
for PHP versions 8.1 and newer.

Also updates the Psalm baseline.

Signed-off-by: Agustin Gomes <me@agustingomes.com>

lcobucci commented on test/Router/IntegrationTest.php:
```diff
@@ -73,9 +74,11 @@ public function createApplicationFromRouter(RouterInterface $router): Applicatio
      * Get the router adapters to test
      *
-     * @psalm-return iterable<class-string>
+     * @psalm-return Generator<int, array{
+     *     0: (AuraRouter::class|FastRouteRouter::class|LaminasRouter::class)
+     * }, mixed, void>
```
I never saw this `(ClassOne::class|ClassTwo::class)` syntax but I think this is more extensible:

```suggestion
     * @psalm-return Generator<int, array{class-string<RouterInterface>}>
```

Also, AFAIK `Generator<int, string>` is a shortcut for `Generator<int, string, mixed, mixed>` and (for this case) we don't really care about types for the generator's input nor its return.

agustingomes replied:
Thanks for the suggestion, I will push that change. 

Initially I changed that when I ran `vendor/bin/psalm --alter --issues=InvalidReturnType`.

lcobucci commented on test/Router/IntegrationTest.php:
```diff
@@ -353,21 +360,25 @@ public function testMatchWithAllHttpMethods(string $adapter, string $method): vo
 
     /**
-     * @psalm-return array<string, array{
+     * @psalm-return iterable<array{
      *     0: class-string,
      *     1: RequestMethod::METHOD_*
      * }>
      */
```
I'd expect us to be more specific here too

```suggestion
     * @psalm-return Generator<int, array{class-string<RouterInterface>, RequestMethod::METHOD_*}>
     */
```

agustingomes replied:
I added the interface to the annotation 👍🏼 

I had to retain the array index in the annotation though, otherwise Psalm would error.

lcobucci commented on test/Router/IntegrationTest.php:
```diff
@@ -353,21 +360,25 @@ public function testMatchWithAllHttpMethods(string $adapter, string $method): vo
+            /**
+             * Excluded from PHP 8.1 onwards since the package seems to be abandoned.
+             */
+            yield [AuraRouter::class, RequestMethod::METHOD_HEAD];
```
We're changing the suite functionality with your changes. We previously had descriptions for the scenarios on this data provider, now that's gone. 

We should have something like this to keep the functionality (for all yields):

```suggestion
            yield 'aura-head' => [AuraRouter::class, RequestMethod::METHOD_HEAD];
```

BTW sorry for not sending all the comments on a single review. It's been a long long day

agustingomes replied:
no worries.

boesing commented on composer.json:
```diff
@@ -30,38 +30,40 @@
+        "laminas/laminas-http": "<2.15.0",
+        "laminas/laminas-stdlib": "<3.6.0",
+        "laminas/laminas-router": "<3.5.0",
+        "filp/whoops": "<2.14.0"
```
I would suggest adding the same constraint here as in `require-dev` as we actually do not test if we are compatible with `2.14.0 - 2.14.3`

weierophinney commented on composer.json:
```diff
@@ -30,38 +30,40 @@
-        "mezzio/mezzio-template": "^2.1.0",
+        "laminas/laminas-httphandlerrunner": "^1.5",
+        "laminas/laminas-stratigility": "^3.5",
+        "laminas/laminas-zendframework-bridge": "^1.4",
```
You can drop the bridge requiremetn entirely, if you make the following additional changes:

- Add a "conflict" rule of `"zendframework/zend-expressive": "*"`
- Remove the "replace" section entirely

weierophinney reviewed (commented):
I've made a suggestion around removing the laminas-zendframework-bridge as part of this patch; I'll push that change to your branch before merging.

Otherwise, :ship:

weierophinney committed:
feat: replace zend-expressive via conflict instead of bridge package

- Adds a conflict rule, marking a conflict against all zend-expressive versions
- Removes the replace rule
- Removes the dependency on the laminas-zendframework-bridge

Signed-off-by: Matthew Weier O'Phinney <matthew@weierophinney.net>
